#4456: Uplift NlpCreateHeads to support generic shapes

TT-BrianLiu opened the pull request:
- Add unit test (harness) for generic NlpCreateHeads
- Split NlpTM into NlpCreateHeads and NlpConcatHeads

TODOs:
- [x] Add second optional input for kv heads
- [x] Clean up and add more unit tests
- [x] Fix broken falcon tests that use `nlp_create_qkv_heads` and make sure no perf regression
- [x] Update docs

Also addressing this issue: https://github.com/tenstorrent-metal/tt-metal/issues/4603

TT-BrianLiu committed:
#4456: Uplift NlpCreateHeads to support generic shapes and optional transpose on K heads
- Add unit test (harness) for generic NlpCreateHeads
- Split NlpTM into NlpCreateHeads and NlpConcatHeads
#4456: Fork old hard-coded CreateHeads for falcon7b
- This might be needed since perf might be worse for the general CreateHeads

TT-BrianLiu committed:
#4456: Update ttnn split_query_key_value_and_split_heads to use new generic nlp_create_qkv_heads

davorchap commented on models/demos/falcon7b/tt/falcon_attention.py:
```diff
@@ -238,7 +238,7 @@ def forward(
         ### TMs ###
         ###########
-        query_layer, key_layer, value_layer = tt_lib.tensor.nlp_create_qkv_heads(
+        query_layer, key_layer, value_layer = tt_lib.tensor.nlp_create_qkv_heads_falcon7b(
```
@TT-BrianLiu and @arakhmati can Falcon we uplifted to use ttnn APIs ?
It's the same OP/kernels under the hood, right?

TT-BrianLiu replied:
I had to change the cb size for the generic implementation. For falcon7b, since head_dim is small, I hard-coded the cb size to fit the entire W dim (double buffered, this is same as BERT i think). For the generic one, I don't base this cb off of input size. It's tiny now, so perf may be worse.

TT-BrianLiu replied:
We should measure and if perf is a lot worse for general shapes, we can optimize the general one further

TT-BrianLiu committed:
#0: Update program cache unit tests to create a dummy tensor before looping
- This shifts the memory so that input/output tensor addresses actually shift
- Otherwise, program callback is not properly tested

TT-BrianLiu commented:
Pipelines:
- post-commit: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7440636452
- fast dispatch: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7440637613
- slow dispatch: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7440637930
